File: built-in plugins dsp/mb-creative-exciter.cpp
/**
 * MB Aural Exciter
 * Category : effect
 * Type     : distortion
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Classic aural exciter for adding sparkle and air
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_CREATIVE_EXCITER_H
#define MB_CREATIVE_EXCITER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCreativeExciter : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-creative-exciter";
    static constexpr const char* PLUGIN_NAME    = "MB Aural Exciter";
    static constexpr const char* PLUGIN_TYPE    = "distortion";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float frequency = 3000f;  // range [500, 10000]
    float harmonics = 0.4f;  // range [0, 1]
    float timbre = 0.5f;  // range [0, 1]
    float mix = 0.3f;  // range [0, 1]
    };

    MbCreativeExciter() = default;
    ~MbCreativeExciter() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.frequency = std::clamp(params.frequency, 500f, 10000f);
        params.harmonics = std::clamp(params.harmonics, 0f, 1f);
        params.timbre = std::clamp(params.timbre, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Aural Exciter
        return input;
    }
};

#endif // MB_CREATIVE_EXCITER_H

File: built-in plugins dsp/mb-creative-exciter.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCreativeExciterPlugin: PluginDefinition = { id: 'mb-creative-exciter', slug: 'mb-creative-exciter', name: 'MB Aural Exciter', category: 'effect', type: 'distortion' as any, version: '1.0.0', description: 'Classic aural exciter for adding sparkle and air', author: 'Max Booster', parameters: [{ id: 'frequency', name: 'Tune', type: 'float', defaultValue: 3000, minValue: 500, maxValue: 10000, automatable: true }, { id: 'harmonics', name: 'Harmonics', type: 'float', defaultValue: 0.4, minValue: 0, maxValue: 1, automatable: true }, { id: 'timbre', name: 'Timbre', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 0.3, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { frequency: 3000, harmonics: 0.4, timbre: 0.5, mix: 0.3 } };

export default MbCreativeExciterPlugin;
